examples/d3d12/src/main.rs

```diff
@@ -0,0 +1,855 @@
+) -> LRESULT {
+    unsafe {
+        let user_data = GetWindowLong(window, GWLP_USERDATA);
+        let sample: Option<&mut S> = match user_data {
+            0 => None,
+            _ => Some(transmute(user_data)),
+        };
+
+        match message {
+            WM_CREATE => {
+                let create_struct: &CREATESTRUCTA = transmute(lparam);
+                SetWindowLong(
+                    window,
+                    GWLP_USERDATA,
+                    transmute(create_struct.lpCreateParams),
+                );
+                LRESULT::default()
+            }
+            WM_DESTROY => {
+                PostQuitMessage(0);
+                LRESULT::default()
+            }
+            _ => {
+                let handled = match sample {
+                    Some(s) => sample_wndproc(s, message, wparam),
+                    None => false,
+                };
+
+                if handled {
+                    LRESULT::default()
+                } else {
+                    DefWindowProcA(window, message, wparam, lparam)
+                }
+            }
+        }
+    }
+}
+
+fn get_hardware_adapter(factory: &IDXGIFactory4) -> Result<IDXGIAdapter1> {
+    let mut adapter_: Option<IDXGIAdapter1> = None;
+
+    let mut i = 0;
+
+    unsafe {
+        while factory.EnumAdapters1(i, &mut adapter_) != DXGI_ERROR_NOT_FOUND {
+            i = i + 1;
+
+            if let Some(adapter) = adapter_ {
+                adapter_ = None;
+                let mut desc: DXGI_ADAPTER_DESC1 = DXGI_ADAPTER_DESC1::default();
+                adapter.GetDesc1(&mut desc).ok()?;
+
+                if (DXGI_ADAPTER_FLAG::from(desc.Flags) & DXGI_ADAPTER_FLAG_SOFTWARE)
+                    != DXGI_ADAPTER_FLAG_NONE
+                {
+                    // Don't select the Basic Render Driver adapter.
+                    // If you want a software adapter, pass in "/warp" on the command line.
+                    continue;
+                }
+
+                // We need the variant where we pass in NULL for the outparam.
+                #[link(name = "d3d12")]
+                extern "system" {
+                    pub fn D3D12CreateDevice(
+                        padapter: ::windows::RawPtr,
+                        minimumfeaturelevel: D3D_FEATURE_LEVEL,
+                        riid: *const ::windows::Guid,
+                        ppdevice: *mut *mut ::std::ffi::c_void,
+                    ) -> ::windows::HRESULT;
+                }
+
+                // Check to see whether the adapter supports Direct3D 12, but
+                // don't create the actual device yet.
+                if D3D12CreateDevice(
+                    adapter.abi(),
+                    D3D_FEATURE_LEVEL_11_0,
+                    &ID3D12Device::IID,
+                    std::ptr::null_mut(),
+                )
+                .is_ok()
+                {
+                    return Ok(adapter);
+                }
+            }
+        }
+    }
```

Comment:
You could save two indent levels here: use `.and_some()?` for the pointer check - afaik `EnumAdapters1` should _always_ return a valid pointer when its result is not `DXGI_ERROR_NOT_FOUND`, because `.and_some` would turn this into `E_POINTER`. The second level is removed by slimming down `unsafe {}` to the three function calls that actually need it. You can also remove the `: Option<IDXGIAdapter1>` and `: DXGI_ADAPTER_DESC1` types, `EnumAdapters1`/`GetDesc1` are already strongly typed.

```suggestion
    for i in 0.. {
        let mut adapter = None;
        let hr = unsafe { factory.EnumAdapters1(i, &mut adapter) };
        if hr == DXGI_ERROR_NOT_FOUND {
            break;
        }

        let adapter = hr.and_some(adapter)?;
        let mut desc = DXGI_ADAPTER_DESC1::default();
        unsafe { adapter.GetDesc1(&mut desc) }.ok()?;

        if (DXGI_ADAPTER_FLAG::from(desc.Flags) & DXGI_ADAPTER_FLAG_SOFTWARE)
            != DXGI_ADAPTER_FLAG_NONE
        {
            // Don't select the Basic Render Driver adapter.
            // If you want a software adapter, pass in "/warp" on the command line.
            continue;
        }

        // We need the variant where we pass in NULL for the outparam.
        #[link(name = "d3d12")]
        extern "system" {
            pub fn D3D12CreateDevice(
                padapter: ::windows::RawPtr,
                minimumfeaturelevel: D3D_FEATURE_LEVEL,
                riid: *const ::windows::Guid,
                ppdevice: *mut *mut ::std::ffi::c_void,
            ) -> ::windows::HRESULT;
        }

        // Check to see whether the adapter supports Direct3D 12, but
        // don't create the actual device yet.
        if unsafe {
            D3D12CreateDevice(
                adapter.abi(),
                D3D_FEATURE_LEVEL_11_0,
                &ID3D12Device::IID,
                std::ptr::null_mut(),
            )
        }
        .is_ok()
        {
            return Ok(adapter);
        }
    }
```

Reply:
Hehe reducing `unsafe` and `mut` are two of the things I look out for while reviewing PRs :)

Reply:
Oh yeah, the compiler can't infer that `for i in 0..` is an infinite iterator... For that you can place `unreachable!()` at the end of the loop.

Anyway, it's not really infinite though, it will run out at some point:

```rust
fn main() {
    for i in (std::u32::MAX - 3).. {
        println!("{}", i);
    }
}
````

```console
me:~/ $ rustc test.sh && ./test
4294967292
4294967293
4294967294
thread 'main' panicked at 'attempt to add with overflow', /home/marijn/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/range.rs:393:1
```